### PR TITLE
Add Compression Utils

### PR DESCRIPTION
**Pull request by SethSmucker**
Add `CompressionAlgorithm` abstract class, used to implement modular compression algo classes.

Created example `GZipCompression : CompressionAlgorithm` as an example.

Added tests for them, although I'm not entirely sure all of the edge cases are covered. These are just what I could find online for GZIP.

Still needs to be implemented in actual classes that use compression.

This is just a proof of concept for feedback.

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/RangeStream.java [hunk 1]
```diff
@@ -204,6 +205,24 @@ public CloseableIterable<QueryPlan> streamPlans(JexlNode script) {
...
+            String serializedTypeMetadata = metadataHelper.getTypeMetadata(config.getDatatypeFilter()).toString();
+            System.out.println("TYPE_METADATA: " + serializedTypeMetadata);
+
+        } catch (TableNotFoundException ignored) {
+
+        }
+
+        // String requiredAuthsString = metadataHelper.getUsersMetadataAuthorizationSubset();
+        // System.out.println("TYPE_METADATA_AUTHS: " + requiredAuthsString);
+
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/RangeStream.java]: This can get removed

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/ShardRangeStream.java [hunk 2]
```diff
@@ -136,18 +136,24 @@ public QueryPlan apply(Entry<Key,Value> entry) {
      * Lift and shift from DefaultQueryPlanner to avoid reliance on static methods
      */
     private void configureTypeMappings(ShardQueryConfiguration config, IteratorSetting cfg, MetadataHelper metadataHelper) {
+        System.out.println("SETH SMUCKER");
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/ShardRangeStream.java]: Prints here can be removed

File: warehouse/core/pom.xml [hunk 3]
```diff
@@ -157,6 +157,12 @@
             <groupId>org.springframework</groupId>
             <artifactId>spring-context-support</artifactId>
         </dependency>
+        <!-- https://mvnrepository.com/artifact/org.tukaani/xz -->
+        <dependency>
+            <groupId>org.tukaani</groupId>
+            <artifactId>xz</artifactId>
+            <version>1.10</version>
+        </dependency>
```
Comment (apmoriarty) [inline warehouse/core/pom.xml:165]: If this is an unused import it can be removed as well

File: warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    /**
+     * GZIP implementation reference: 1) String → bytes via charset 2) Write to GZIPOutputStream (must close to finalize) 3) Base64-encode compressed bytes
+     */
+    private String compressGZIP(final String data, final Charset charset) {
+        final byte[] input = data.getBytes(charset);
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream(); GZIPOutputStream gzip = new GZIPOutputStream(baos)) {
+
+            gzip.write(input);
+            gzip.close(); // must close to flush all compressed data into baos
```
Comment (apmoriarty) [inline warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java]: If you need the gzip stream to e closed you can always nest another try-with-resources block. Same comment goes for other instances in this class

File: warehouse/core/pom.xml (hunk 3 above)
Comment (SethSmucker) [reply warehouse/core/pom.xml:165]: This is used for Apache Commons ZLMACompressorOutputStream

File: warehouse/query-core/src/main/java/datawave/query/index/lookup/RangeStream.java [hunk 5]
```diff
@@ -204,6 +205,10 @@ public CloseableIterable<QueryPlan> streamPlans(JexlNode script) {
             log.trace(JexlStringBuildingVisitor.buildQuery(node));
         }
 
+        Multimap<String,Type<?>> nonIndexedQueryFieldsDatatypes = HashMultimap.create(config.getQueryFieldsDatatypes());
+        nonIndexedQueryFieldsDatatypes.keySet().removeAll(config.getIndexedFields());
+        String nonIndexedTypes = QueryOptions.buildFieldNormalizerString(nonIndexedQueryFieldsDatatypes);
+
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/index/lookup/RangeStream.java]: Just one more extra to remove

File: warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    /**
+     * "7-Zip" via LZMA stream implementation: 1) String → bytes via charset 2) Write to LZMACompressorOutputStream (close to finalize) 3) Base64-encode
+     * compressed bytes
+     *
+     * Note: This uses the LZMA compressor stream from Apache Commons Compress, which is the algorithm used by 7-Zip; it is not a .7z archive container.
+     */
+    private String compress7ZIP(final String data, final Charset charset) {
+        final byte[] input = data.getBytes(charset);
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
```
Comment (avgAGB) [inline warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java]: just wondering, could the stream be initialized based off of the length of the input array above, sort of like what happens on line 129?

File: warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        }
+    }
+
+    /**
+     * GZIP implementation reference: 1) String → bytes via charset 2) Write to GZIPOutputStream (must close to finalize) 3) Base64-encode compressed bytes
+     */
+    private String compressGZIP(final String data, final Charset charset) {
+        final byte[] input = data.getBytes(charset);
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
```
Comment (avgAGB) [inline warehouse/core/src/main/java/datawave/util/compression/OptionCompressor.java]: other spot to possibly initialize with an explicit value
